src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -614,6 +614,77 @@ QString QgsPostgresUtils::whereClause( const QgsFeatureIds &featureIds, const Qg
     return expr;
   };
 
+  auto lookupComboKeyWhereClause = [ = ]( bool canUseVALUES, QList<QString> withCastTypeName )
+  {
+    if ( pkAttrs.size() == 1 )
+      return lookupKeyWhereClause();
+
+    if ( featureIds.isEmpty() )
+      return QString();
+
+    QString delim;
+    QString whereKeys;
+    for ( int i = 0; i < pkAttrs.size(); i++ )
+    {
+      whereKeys += delim + QgsPostgresConn::quotedIdentifier( fields.at( pkAttrs[i] ).name() );
+      delim = QStringLiteral( "," );
+    }
+
+    QgsFeatureIds fidsToOr;
+    QStringList whereValuesList;
+    for ( const QgsFeatureId featureId : std::as_const( featureIds ) )
+    {
+      const QVariantList pkVals = sharedData->lookupKey( featureId );
+      if ( !pkVals.isEmpty() )
+      {
+        Q_ASSERT( pkVals.size() == pkAttrs.size() );
+
+        QString delim;
+        QString whereValues;
+        bool isNotNull = true;
+        for ( int i = 0; i < pkAttrs.size(); i++ )
+        {
+          if ( pkVals[i].isNull() )
+            isNotNull = false;
+          whereValues += delim + QgsPostgresConn::quotedValue( pkVals[i] );
+          if ( canUseVALUES && withCastTypeName.contains( fields.at( pkAttrs[i] ).typeName() ) )
+            whereValues += QStringLiteral( "::%1" ).arg( fields.at( pkAttrs[i] ).typeName() );
+          delim = QStringLiteral( "," );
+        }
+        if ( isNotNull )
+          whereValuesList << whereValues;
+        else
+          fidsToOr.insert( featureId );
```

Comment:
Is it possible to have a composite primary key with one of value equals to NULL ? Is it not a bad database design pattern?

And so there is a different behavior between IN and OR ? 

IN (1, NULL) returns no rows while a == 1 AND b == NULL returns a row ?

Reply:
> Is it possible to have a composite primary key with one of value equals to NULL ? Is it not a bad database design pattern?

I agree that using columns that can contain NULLs as a PK is a very bad practice. But the original code supports this feature, so I had to implement this feature in this PR.

> And so there is a different behavior between IN and OR ?
> 
> IN (1, NULL) returns no rows while a == 1 AND b == NULL returns a row ?

not quite so, the original code creates a construction a = 1 AND b is NULL

Reply:
> I agree that using columns that can contain NULLs as a PK is a very bad practice

Well, I tested it and it's not even possible

```sql
composite_null=# create table testcomp ( id1 integer, id2 integer, PRIMARY KEY (id1, id2 ) ); 
CREATE TABLE
composite_null=# \d
           Liste des relations
 Schéma |   Nom    | Type  | Propriétaire 
--------+----------+-------+--------------
 public | testcomp | table | julien
(1 ligne)

composite_null=# insert into testcomp ("id1", "id2") VALUES(1,2); 
INSERT 0 1
composite_null=# insert into testcomp ("id1", "id2") VALUES(NULL,2); 
ERREUR:  une valeur NULL viole la contrainte NOT NULL de la colonne « id1 »
DÉTAIL : La ligne en échec contient (null, 2).

```

So I think it will be better to remove all the OR specific code and display a error message if one pkVal is null

Reply:
...or maybe just print a debug message and just keep the IN clause

Reply:
Here I have to disagree with you. It's hard for me to predict all the consequences of such changes.
But if the goal is to prohibit the use of NULL in the multi-column PK, then this should be done at the stage of adding such layers to QGIS. Perhaps at the stage of testing the PK for uniqueness. Otherwise, everything will look extremely illogical, the user will be able to add a layer to QGIS, work with it... suddenly, some tools will begin to shower with warnings that he is doing everything wrong.

Reply:
I agree with you it's not the place to prohibit the use of null, forget about the error message. My goal here is to simplify the code to not deal with OR inside the lambda and stick with only IN clause. 

From what I understand, the behavior with IN when there is some NULL in composite pkeys is the same as OR. So I would prefer to remove the OR part inside the lambda. 

This PR propose to use IN for certain type in composite pey, so why using OR when the value is NULL?

Reply:
> My goal here is to simplify the code to not deal with OR inside the lambda and stick with only IN clause.
> 
> From what I understand, the behavior with IN when there is some NULL in composite pkeys is the same as OR. So I would prefer to remove the OR part inside the lambda.
> 
> This PR propose to use IN for certain type in composite pey, so why using OR when the value is NULL?

In my opinion, this will lead to unnecessary complication of the lambda function code.

Reply:
Why complications? I propose to remove all things in lambda function related to OR and keep only IN inside

Reply:
> Why complications? I propose to remove all things in lambda function related to OR and keep only IN inside

What for? What is the purpose of these changes?
Unfortunately, I can't understand what changes you propose to make :(

Reply:
The `lookupKeyWhereClause` method was initially designed to be called when we choose to use IN clause as the returned whereClause. I propose to let it that way, meaning that this method doesn't return any OR clause, and return only IN clause.

In other words I propose to remove whereOrList, whereOrstring variables and the related code.

Reply:
Done